### ISSUE TITLE
internal error with lambda (x): ...

### ISSUE THREAD

**Issue by michal-au** [python/mypy#1312]
test.py:
```filter(lambda (x): x > 0, [0,1])```

```
...$ mypy --py2 test.py 
Traceback (most recent call last):
  File "/usr/local/bin/mypy", line 6, in <module>
    main(__file__)
  File "/usr/local/lib/python3.4/dist-packages/mypy/main.py", line 54, in main
    type_check_only(sources, bin_dir, options)
  File "/usr/local/lib/python3.4/dist-packages/mypy/main.py", line 98, in type_check_only
    python_path=options.python_path)
  File "/usr/local/lib/python3.4/dist-packages/mypy/build.py", line 206, in build
    result = manager.process(initial_states)
  File "/usr/local/lib/python3.4/dist-packages/mypy/build.py", line 403, in process
    next.process()
  File "/usr/local/lib/python3.4/dist-packages/mypy/build.py", line 907, in process
    self.type_checker().visit_file(self.tree, self.tree.path)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 406, in visit_file
    self.accept(d)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 447, in accept
    typ = node.accept(self)
  File "/usr/local/lib/python3.4/dist-packages/mypy/nodes.py", line 540, in accept
    return visitor.visit_expression_stmt(self)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 1534, in visit_expression_stmt
    self.accept(s.expr)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 447, in accept
    typ = node.accept(self)
  File "/usr/local/lib/python3.4/dist-packages/mypy/nodes.py", line 965, in accept
    return visitor.visit_call_expr(self)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 1946, in visit_call_expr
    return self.expr_checker.visit_call_expr(e)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checkexpr.py", line 142, in visit_call_expr
    return self.check_call_expr_with_callee_type(callee_type, e)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checkexpr.py", line 193, in check_call_expr_with_callee_type
    e.arg_names, callable_node=e.callee)[0]
  File "/usr/local/lib/python3.4/dist-packages/mypy/checkexpr.py", line 233, in check_call
    callee, args, arg_kinds, formal_to_actual, context)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checkexpr.py", line 399, in infer_function_type_arguments
    callee_type, args, arg_kinds, formal_to_actual)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checkexpr.py", line 328, in infer_arg_types_in_context2
    res[ai] = self.accept(args[ai], callee.arg_types[i])
  File "/usr/local/lib/python3.4/dist-packages/mypy/checkexpr.py", line 1423, in accept
    return self.chk.accept(node, context)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 447, in accept
    typ = node.accept(self)
  File "/usr/local/lib/python3.4/dist-packages/mypy/nodes.py", line 1187, in accept
    return visitor.visit_func_expr(self)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 2071, in visit_func_expr
    return self.expr_checker.visit_func_expr(e)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checkexpr.py", line 1249, in visit_func_expr
    self.chk.check_func_item(e, type_override=inferred_type)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 620, in check_func_item
    self.check_func_def(defn, typ, name)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 717, in check_func_def
    self.accept_in_frame(item.body)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 463, in accept_in_frame
    changed, _ = self.binder.pop_frame(True, True)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 209, in pop_frame
    changed = self.update_from_options(options)
  File "/usr/local/lib/python3.4/dist-packages/mypy/checker.py", line 161, in update_from_options
    if not is_same_type(type, current_value):
  File "/usr/local/lib/python3.4/dist-packages/mypy/sametypes.py", line 27, in is_same_type
    return left.accept(SameTypeVisitor(right))
  File "/usr/local/lib/python3.4/dist-packages/mypy/types.py", line 157, in accept
    return visitor.visit_erased_type(self)
  File "/usr/local/lib/python3.4/dist-packages/mypy/sametypes.py", line 74, in visit_erased_type
    raise RuntimeError()
RuntimeError: 

*** INTERNAL ERROR ***

test.py:1: error: Internal error -- please report a bug at https://github.com/JukkaL/mypy/issues
```

**Comment by gvanrossum** [comment]
Thanks for reporting this! It's almost certainly due to the way parameter unpacking treats this end case. Do you feel up to trying to come up with a fix yourself?

**Comment by michal-au** [comment]
Sure, I can give it a try... We'll see what happens.

**Comment by gvanrossum** [comment]
Reid, does this still feel like something you'd like to do? I can also take a look myself.

**Comment by rwbarton** [comment]
I would be happy to have you look at it; actually I don't even know what `lambda (x): ...` is supposed to mean in the first place :)

**Comment by gvanrossum** [comment]
OK. Quite possibly I assigned it to you by mistake.

FWIW `lambda (x): ...` is the same as `lambda x: ...` -- it's an edge case of `lambda (x, y): ...` which is Python 2 only syntax that requires a single argument that's a tuple of 2 values.

**Comment by gvanrossum** [comment]
Closed by b4b0c1c70670c3064f690665af4d71f447ba00be (PR #1384).
